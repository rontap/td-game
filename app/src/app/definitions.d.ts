interface Map {
    map(params: any): any[];
}

