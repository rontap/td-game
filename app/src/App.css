@import url('https://fonts.cdnfonts.com/css/bahnschrift');


.App {
    text-align: center;
}

.App-logo {
    height: 40vmin;
    pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
    .App-logo {
        animation: App-logo-spin infinite 20s linear;
    }
}

.App-header {
    background-color: #282c34;
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    font-size: calc(10px + 2vmin);
    color: white;
}

.App-link {
    color: #61dafb;
}

@keyframes App-logo-spin {
    from {
        transform: rotate(0deg);
    }
    to {
        transform: rotate(360deg);
    }
}

.svgRoot {
    /*border: 3px double whitesmoke;*/
    border-radius: 10px;

}


.svgRoot .static {
    cursor: not-allowed;
}

.svgRoot .draggable {
    cursor: move;
}

.svgRoot .draggable:hover {
    fill: orange;
}

.svgBlueprint {
    position: fixed;
    left: 350px;
    top: 135px;

}

.boxedItem {
    border: 2px solid #ccc;
    border-radius: 5px;
    width: 100px;
    height: 60px;
}

foreignObject {
    outline: 3px solid transparent;

}

foreignObject:hover {
    outline: 3px double whitesmoke;
}


foreignObject > div {
    background: #66666666;
}

svg:has(foreignObject:active div) {
    cursor: move;
}

.blue div {
    border-color: #4488FF;
}

.red div {
    border-color: #e91e63;
}

.orange div {
    border-color: #ff693e;
}

.green div {
    border-color: #4caf50;
}

.purple div {
    border-color: #9c27b0;
}


.boxedItem .title {
    font-size: 15px;
    padding: 3px;
    margin-bottom: 3px;
}

.orange .boxedItem .title {
    background: #ff693e;
}

.blue .boxedItem .title {
    background: #4488FF;
}

.red .boxedItem .title {
    background: #e91e63;
}


.green .boxedItem .title {
    background: #4caf50;
}

.purple .boxedItem .title {
    background: #9c27b0;
}

#activeNodes {
    position: fixed;
    top: calc(20px + 400px + 80px);
    height: 400px;
    width: 320px;
}

#availableNodes {
    position: fixed;
    top: 80px;
    height: 400px;
    width: 320px;

}

#availableNodes .btnVertical .btn {
    display: block;
}

#propertyRoot, #activeNodes, nav, .majorElement {
    position: fixed;

    background: #263136;
    border-radius: 10px;
    margin: 15px;
    border: 1px solid whitesmoke;

}

#nodeBlueprints {
    width: 300px;
    top: 120px;
    height: 600px;
    text-align: left;
    padding: 15px;
}

#nodeBlueprints .nodeListItem {
    margin: 15px 0;
    display: block
}

#zoomCtn {
    bottom: 0px;
    left: 0px;

}

[center], .center {
    text-align: center;
}

button.newBtn.btn {
    text-align: center;
    width: 150px;
    margin-left: 10px;
    display: block;
}

nav {
    top: 0;
    width: max(40%, 800px);
    left: 0;
    padding: 10px;
    text-align: left;
}

#header {


}

nav #titlemark {
    font-size: 26px;
    color: whitesmoke;
    font-weight: bold;
    margin: 0 15px 0 10px;
    vertical-align: -8px;


    text-align: right;
    display: inline-block;
    background: whitesmoke;
    background: -webkit-linear-gradient(0deg, whitesmoke, #d4effc);
    -webkit-background-clip: text;
    -webkit-text-fill-color: transparent;
}

#propertyRoot {
    top: 0px;
    min-height: 400px;
    right: 0;
    width: 310px;
}

body {
    background: #263238;
    color: #eee;
    /*transition: color .8s , background .8s;*/
}


.btn.btn-100 {
    min-width: 100px;
    display: inline-block;
}

#svgRootCont {
    margin-top: 0px;

}

body {
    overflow: hidden;
}

.no-border {
    border: none;
}

.nodeListItem {


    margin: 5px 5px;
}

.nodeListItem .listItemDescr {
    width: 60px;
    display: inline-block;
}

#activeNodes {
    overflow: auto;
    max-height: 400px;
}

rect:active {
    cursor: move;
}

button, h3, #titlemark {
    font-family: Bahnschrift, sans-serif;
}

#smallGrid path {
    transition: stroke-width .5s;
}

svg:has(rect:active) #smallGrid path {
    stroke-width: 0px;
}

svg:has(rect:active) .nodeConnection {
    opacity: 0;
}


svg:not(:has(.currentNodeActive)) .nodeConnectionStart {
    opacity: 0;
}

svg:has(.currentNodeActive) .nodeConnectionEnd {
    opacity: 0;
}

.currentBez {
    stroke: orange;
    stroke-width: 2;
    fill: transparent;
}

.nodeConnection {
    border: 0px solid #999;
    width: 10px;
    height: 16px;
    padding: 0;
    border-radius: 0 4px 4px 0;
    position: absolute;
    top: 23px;
    z-index: 3;
    background: #ccc;
    transition: opacity .1s;
    cursor: pointer;
}

.nodeConnectionStart {
    left: 0px;
    border-radius: 0 4px 4px 0;
}

.nodeConnectionEnd {
    right: 0px;
    border-radius: 4px 0px 0px 4px;
}

.nodeConnection:hover {
    background: #4caf50;
}

.closer {
    float: right;
}

#recenter {
    left: calc(50% - (81px / 2));
    bottom: -3px;
}

.grid {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-rows: 1fr;
    gap: 0px 0px;
    grid-template-areas:
    ". .";
}

.gtc-2 {
    grid-template-columns: 50% 50%;
    grid-template-rows: 1fr;
}

.majorElement hr {
    border: none;
    height: 1px;
    background: whitesmoke;
}

.dnodes .btn {
    width: calc(50% - 20px);
}
.dnodes .btn {
    width: calc(33% - 10px);
}

.w-100 {
    width: 100%;
}
.gtc-m-10 .gridItem {
    margin:10px;
}