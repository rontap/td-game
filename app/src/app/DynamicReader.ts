import {jsobj} from "./util";

const readPath = '../dynamic/node';

export interface TypeList {
    [key: string]: string[] // adjusting require this in order to some json data type
}

export function loadJsonNodeDefinitions(): NTypeMap {
    const context = require.context('../dynamic/nodes/', true, /\.(json)$/);
    let files = new Map<string, jsobj>();
    context.keys().forEach((filename) => {
        files.set(filename, context(filename));
    });
    return files;
}

export type NTypeMap = Map<string, jsobj>;