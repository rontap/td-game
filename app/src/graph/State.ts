import {jsobj} from "../app/util";
import {create, useStore} from 'zustand'
import {devtools, persist} from 'zustand/middleware'
import {Node} from "../node/Node";
import {Line, LineId, NodeId} from "../node/Line";
import {unwatchFile} from "fs";
import {temporal, TemporalState} from 'zundo'
import {shallow} from "zustand/shallow";
import {GraphUtil, GraphUtilInst} from "./GraphUtil";

// export default class State {
//     static nodes: Node[] = [];
// }

interface AppState {
    nodes: Node[],
    lines: Line[],
    addNode: (node: Node) => void
    getNodeById: (id: number) => Node | undefined
    removeNode: (id: number) => void,
    addLine: (line: Line) => boolean,
    removeLine: (id: number) => void,

    resetStore: () => void,

    blueprintedNode: string,
    getLineBetween: (from: number, to: number) => Line | undefined
    zoom: number,
    contextMenu: any,
    lineAddAt: Node | undefined,
    activeNode: Node | undefined,
    setActiveNode: (id?: NodeId | undefined) => void,
    getLinesAtNodeConnection: (id: NodeId | undefined, end: End) => Line[],
    setBlueprintedNode: (nodeName: string) => void,
    forceSvgRender: {},
    doSvgRender: () => void,

    setSingleEdgeOfActiveNode:
        (propertyName: string, to: string) => void,
}

export enum End { FROM, TO}

const initialState = {
    lines: [],
    lineAddAt: undefined,
    contextMenu: {},
    blueprintedNode: "",
    activeNode: undefined,
    forceSvgRender: {},
    nodes: [],
    zoom: 1,
}

const actions = {}

const State = create<AppState>()(
    devtools(
        temporal(
            //persist(
            (set, get) => ({
                setSingleEdgeOfActiveNode: (propertyName: string, to: string) =>
                    set((state) => {
                            const activeNode = state.activeNode;
                            console.log('->>', activeNode, to);
                            if (!activeNode) return {};
                            activeNode._configurableInputValues.set(propertyName, to);
                            return {activeNode}
                        }
                    ),
                addNode: (node: Node) => set((state) =>
                    ({nodes: state.nodes.concat(node)})),
                getNodeById: (id: NodeId) =>
                    get().nodes.find(item => item.ID === Number(id)),
                removeNode: (id: NodeId) => set((state) =>
                    ({nodes: state.nodes.filter(item => item.ID !== Number(id))})),
                addLine: (line: Line) => {
                    if (get().getLineBetween(line.from, line.to)) {
                        return false;
                    }
                    set((state) => ({lines: state.lines.concat(line)}));
                    GraphUtilInst.detectCircles();
                    return true;
                },
                removeLine: (id: LineId) => {
                    set((state) => ({lines: state.lines.filter(item => item.ID !== Number(id))}));
                    GraphUtilInst.detectCircles();
                },
                getLineBetween: (from: number, to: number) =>
                    get().lines.find(line => line.from === from && line.to === to),
                setActiveNode: (id) => set((state) =>
                    ({activeNode: id ? get().nodes.find(item => item.ID === Number(id)) : undefined})
                ),
                setBlueprintedNode: (nodeName: string) => set((state) =>
                    ({blueprintedNode: nodeName})),
                getLinesAtNodeConnection: (id: NodeId | undefined, end: End) => {
                    if (!id) return [];
                    const whichJunction = end === End.FROM ? "from" : "to";
                    return get().lines.filter(line => line[whichJunction] === id)
                },
                doSvgRender: () => set((state) =>
                    ({forceSvgRender: {}})
                ),
                resetStore: () => set((state) => initialState),

                // store part
                ...initialState
            }),
            {
                partialize: (state) => {
                    const {lines, nodes, ...rest} = state
                    return {lines, nodes}
                },
                equality: shallow
            }
            //)
        ) // temporal
        , {
            name: 'store',
        }
    )
);

const useTemporalStore = <T, >(
    selector: (state: TemporalState<any>) => T,
    equality?: (a: T, b: T) => boolean,
) => useStore(State.temporal, selector, equality);

export default State;
const {getState, setState, subscribe} = State
export {
    getState,
    setState,
    subscribe,
    useTemporalStore
};
// @ts-ignore
window.State = State;