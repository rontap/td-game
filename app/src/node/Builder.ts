import {loadJsonNodeDefinitions, NodeTemplate, NodeTemplateMap} from "../app/DynamicReader";
import {EdgeInvariant, EdgeLoader, edgeTypes} from "../graph/EdgeLoader";
import {jsobj} from "../app/util";
import {Node} from './Node';
import State from "../graph/State";

EdgeLoader();

export class NodeBuilder {
    static get rawTypes(): NodeTemplateMap {
        return this._rawTypes;
    }

    static get types(): NodeTemplateMap {
        return this._types;
    }

    private static _rawTypes = new Map<string, NodeTemplate>();
    private static _types = new Map<string, NodeTemplate>();

    static Build() {
        this._rawTypes = loadJsonNodeDefinitions();

        [...this._rawTypes.values()].map((value: NodeTemplate) => {
            this._types.set(value.name, value);
        })

        return this._rawTypes;
    }

    static New(nodeType: string) {
        const nodes = State.getState().nodes;
        State.setState({nodes: nodes.concat(new Node(nodeType))})
    }

    static getType(name: string) {
        return this.types.get(name);
    }

    // static InstantiateNode(nodeTemplate: jsobj) {
    //     switch (nodeTemplate.type) {
    //         case "default":
    //             return new Node(nodeTemplate, nodeTemplate.type);
    //         case undefined:
    //             throw Error("Node type property is required for instantiation.");
    //         default:
    //             throw Error("Unknown Node type: " + nodeTemplate.type);
    //     }
    // }

}

EdgeInvariant(true);