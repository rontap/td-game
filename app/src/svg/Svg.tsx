import CONST from "../const";
import Draggable, {DragHandler, DragHandlerInst, Geom, Point} from "./Draggable";
import {useEffect} from "react";
import {jsobj} from "../app/util";
import State, {getState} from "../graph/State";

import {Node} from "../node/Node";
import Movable from "./Movable.js";
import {Line} from "../node/Line";
import SvgLines from "./SvgLines";

export default function Svg(props: jsobj) {
    const {blueprint} = props;
    useEffect(() => {
        DragHandlerInst.getTransformMatrix();
    }, []);

    let nodes, lines, tempSvgRender;


    console.log('rendering svg')
    nodes = State((state) => state.nodes)


    return (
        <div id={"svgRootCont"}
             {...Movable}
        >
            <svg
                {...Draggable}
                {...(blueprint ? CONST.blueprintRectSize : CONST.rectSize)}
                className={"svgRoot"}
            >

                <defs>
                    <pattern id="smallGrid" width="8" height="8" patternUnits="userSpaceOnUse">
                        <path d="M 8 0 L 0 0 0 8" fill="none" stroke="#384850" strokeWidth="0.5"/>
                    </pattern>
                    <pattern id="grid" width="80" height="80" patternUnits="userSpaceOnUse">
                        <rect width="80" height="80" fill="url(#smallGrid)"/>
                        <path d="M 80 0 L 0 0 0 80" fill="none" stroke="#5b6d76" strokeWidth="1"/>
                    </pattern>
                </defs>


                <defs>
                    <marker
                        id="dot"
                        viewBox="0 0 10 10"
                        refX="5"
                        refY="5"
                        markerWidth="5"
                        markerHeight="5">
                        <circle cx="5" cy="20" r="20" fill="red"/>
                        {/*<animateMotion*/}

                        {/*    dur="3s"*/}

                        {/*    repeatCount="indefinite"*/}
                        {/*    path={Geom.bezierSvgD(fromPoint, toPoint)}/>*/}

                    </marker>
                </defs>

                <rect
                    x="-100%"
                    y="-100%"
                    width="1000%"
                    height="10000%" fill="url(#grid)"/>

                {/*<rect className="draggable f2" x="160" y="190" width="80" height="40"*/}

                {/*      fill="#5577ff">*/}
                {/*</rect>*/}

                {/*<rect className="draggable" x="350" y="210" width="80" height="40"*/}

                {/*      fill="#ff77aa">*/}
                {/*</rect>*/}


                {
                    getCurrentLine()
                }

                <SvgLines/>


                {
                    nodes.map((node: Node) => node.getSvg())
                }

                {/*{*/}
                {/*    nodes.map((node: Node) => node.getInputLines())*/}
                {/*}*/}


                {/*<path d="M100,100 C250,100 250,250 400,250"*/}
                {/*      style={{fill: 'transparent', stroke: 'red'}}/>*/}

                {/*<line*/}
                {/*    className={"data-line-0-1"}*/}
                {/*    x1="0" y1="80" x2="100" y2="20" stroke="black"/>*/}

                {/*<foreignObject x="200" y="20" width="160" height="100">*/}
                {/*    <div className={"boxedItem"}>*/}
                {/*        hi<br/>*/}
                {/*        <button>btn</button>*/}
                {/*    </div>*/}
                {/*</foreignObject>*/}


                {/*<defs>*/}

                {/*    <marker*/}
                {/*        id="dot"*/}
                {/*        viewBox="0 0 10 10"*/}
                {/*        refX="5"*/}
                {/*        refY="5"*/}
                {/*        markerWidth="5"*/}
                {/*        markerHeight="5">*/}
                {/*        <circle cx="5" cy="5" r="5" fill="red"/>*/}
                {/*    </marker>*/}
                {/*</defs>*/}

                {/*<polyline*/}
                {/*    points="15,80 29,50 43,60 57,30 71,40 85,15"*/}
                {/*    fill="none"*/}
                {/*    stroke="grey"*/}
                {/*    markerStart="url(#dot)"*/}
                {/*    markerMid="url(#dot)"*/}
                {/*    markerEnd="url(#dot)"/>*/}

                {/*<path*/}
                {/*    d="M30.367029359882025,222.74918426476106 C132.1095737779142,222.74918426476106 132.1095737779142,104.44504125761989 233.85211819594636,104.44504125761989"*/}
                {/*    className="data-curve-from-2 data-curve-to-1" x1="30.367029359882025" y1="222.74918426476106"*/}
                {/*    x2="233.85212819594636" y2="104.4450512576199"*/}
                {/*    markerStart="url(#dot)"*/}
                {/*    markerMid="url(#dot)"*/}
                {/*    markerEnd="url(#dot)"*/}
                {/*    style={{fill: "transparent", stroke: "whitesmoke", width: "3px"}}*/}
                {/*    path="M30.367029359882025,222.74918426476106 C132.1095737779142,222.74918426476106 132.1095737779142,104.44504125761989 233.85211819594636,104.44504125761989"></path>*/}

            </svg>
        </div>
    );
}
const getCurrentLine = () => {
    const lineAddAt = State((state) => state.lineAddAt)
    const node = State((state) => state.getNodeById(lineAddAt?.id || -1))

    if (!node?.coords) {
        return <></>
    }

    if (!lineAddAt.id) {
        return <></>
    }

    const fromPoint = DragHandler.getCoords(node.selfSvg).add(CONST.box.width, CONST.box.pointTop);
    const toPoint = DragHandlerInst.getCursor(lineAddAt.evt);

    return <>
        <path d={Geom.bezierSvgD(fromPoint, toPoint)}
              className={"currentBez currentNodeActive"}/>
    </>
}