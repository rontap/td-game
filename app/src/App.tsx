import React, {useState} from 'react';

//---------------------
import './ui/styles/App.css';
import './ui/styles/svg.css';
import './ui/styles/Stem.css';
import './ui/styles/ctxmenu.css';
import './ui/styles/mat.css';

import {NodeBuilder} from "./node/Builder";
import Svg from "./svg/Svg";
import {jsobj} from './util/util';
import ZoomInfo from "./ui/ZoomInfo";
import State from "./graph/State";
import Button from "./ui/components/Button";
import ContextMenu from "./ui/components/ContextMenu";
import Header from "./ui/Header";
import NodeBlueprints, {NodeGroups} from "./ui/NodeBlueprints";
import BlueprintSvg from "./svg/BlueprintSvg";
import NodeBlueprintConfigEditor from "./node/NodeBlueprintConfigEditor";
import Taskbar from "./ui/Taskbar";


function App() {
    const ng = State((state) => state.nodeGroup)
    const items: Map<string, jsobj> = NodeBuilder.Rebuild();

    const [graph, setGraph] = useState(true);

    const [light, setLight] = useState(localStorage.getItem("graphene_theme") === "1");
    const [min, setMin] = useState(false);
    const toggleBg = () => {
        setLight(now => !now);
    };
    const toggleMin = () => {
        setMin(now => !now);
    }

    return (
        <div className={`App ${light ? "_white" : ""} ${min ? "_min" : ""}`}>
            <nav>
                <span id={"titlemark"}>GRAPHENE</span>
                <Button
                    className={"blue"}
                    disabled={!graph}
                    onClick={() => setGraph(false)}>Group</Button>
                <Button
                    disabled={graph}
                    className={"blue"}
                    onClick={() => setGraph(true)}>Graph</Button>
                <Header toggleBg={toggleBg}
                        toggleMin={toggleMin}
                        graph={graph}/>
                <a id={"vers"}
                   target={"_blank"}
                   href={"https://github.com/rontap/thesis"}
                   title={"Github Source"}
                >1.0 Thesis release</a>

            </nav>

            {graph ? <>
                    <ContextMenu items={items}/>
                    <ZoomInfo/>
                    <Svg items={items}/>
                    <Taskbar items={items}/>
                </>
                : <>
                    <div id={"groupsCtnr"}>
                        <NodeGroups items={items}/>
                        <NodeBlueprints items={items}/>
                        <NodeBlueprintConfigEditor items={items}/>
                        <BlueprintSvg items={items} blueprint/>
                    </div>
                </>}


        </div>
    );
}

export default App;
