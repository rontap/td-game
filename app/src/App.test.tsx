import Svg from "./svg/Svg";
import {NodeBuilder} from "./node/Builder";
// @ts-ignore
import renderer from "react-test-renderer";
import App from "./App";
import {getState} from "./graph/State";
import {NodeGroup} from "./app/DynamicReader";

it('Renders <App/> without issues', () => {
    getState().resetStore();
    NodeGroup.loadDebugNodes = true;
    const appCore = renderer.create(<App/>)
        .toJSON();
    expect(appCore).toMatchSnapshot();
})