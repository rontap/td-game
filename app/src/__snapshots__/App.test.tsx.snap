// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <App/> without issues 1`] = `
<div
  className="App "
>
  <nav>
    <span
      id="titlemark"
    >
      GRAPHENE
    </span>
    <button
      className=" btn blue"
      disabled={false}
      onClick={[Function]}
    >
      Edit Group
    </button>
    <button
      className=" btn blue"
      disabled={true}
      onClick={[Function]}
    >
      Edit Graph
    </button>
    <span
      id="header"
    >
      <div
        className="btnGroup undefined"
      >
        <button
          className=" btn "
          onClick={[Function]}
        >
          Undo
        </button>
        <button
          className=" btn "
          onClick={[Function]}
        >
          Redo
        </button>
      </div>
      <button
        className=" btn blue"
      >
        Run
      </button>
        
      <button
        className=" btn "
        onClick={[Function]}
      >
        Auto Layout
      </button>
       
      <button
        className=" btn "
        onClick={[Function]}
      >
        Toggle Theme
      </button>
      <button
        className=" btn "
        onClick={[Function]}
      >
        FRF
      </button>
    </span>
  </nav>
  <div
    className="hiddenCtx"
    id="ctxMenu"
    style={
      Object {
        "top": "-10px",
      }
    }
  />
  <div
    className="majorElement no-border"
    id="zoomCtn"
  >
    <div
      className="btnGroup undefined"
    >
      <button
        className=" btn "
        disabled={false}
        onClick={[Function]}
      >
        -
      </button>
      <button
        className=" btn "
        disabled={true}
        onClick={[Function]}
        title="Reset zoom"
      >
        Zoom 
        100
        %
      </button>
      <button
        className=" btn "
        disabled={false}
        onClick={[Function]}
      >
        +
      </button>
    </div>
  </div>
  <div
    id="svgRootCont"
    onClick={[Function]}
    onContextMenu={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onWheel={[Function]}
  >
    <svg
      className="svgRoot"
      clientHeight={768}
      clientWidth={1024}
      height={768}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      width={1024}
      x={0}
      xmlns="http://www.w3.org/2000/svg"
      y={0}
    >
      <defs>
        <pattern
          height="8"
          id="smallGrid"
          patternUnits="userSpaceOnUse"
          width="8"
        >
          <path
            d="M 8 0 L 0 0 0 8"
            fill="none"
            stroke="#384850"
            strokeWidth="0.5"
          />
        </pattern>
        <pattern
          height="80"
          id="grid"
          patternUnits="userSpaceOnUse"
          width="80"
        >
          <rect
            fill="url(#smallGrid)"
            height="80"
            width="80"
          />
          <path
            d="M 80 0 L 0 0 0 80"
            fill="none"
            stroke="#5b6d76"
            strokeWidth="1"
          />
        </pattern>
      </defs>
      <defs>
        <marker
          id="dot"
          markerHeight="5"
          markerWidth="5"
          refX="5"
          refY="5"
          viewBox="0 0 10 10"
        >
          <circle
            cx="5"
            cy="20"
            fill="red"
            r="20"
          />
        </marker>
      </defs>
      <rect
        fill="url(#grid)"
        height="10000%"
        id="bgRectSvg"
        width="1000%"
        x="-100%"
        y="-100%"
      />
    </svg>
  </div>
  <div
    id="taskbar"
  >
    <div
      className="taskbarElement"
      style={
        Object {
          "height": "200px",
          "transform": "translate3d(0px, 0px, 0)",
          "width": "350px",
        }
      }
    >
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled={false}
        aria-roledescription="draggable"
        className="taskbarHeader"
        onKeyDown={[Function]}
        onPointerDown={[Function]}
        role="button"
        tabIndex={0}
        title="Drag Element"
      />
      <button
        className="minimiseButton"
        onClick={[Function]}
      >
        _
      </button>
      <div>
        <div
          id="availableNodes"
        >
          <h3>
            Add Node
          </h3>
          <div
            className="dnodes"
          >
            <button
              className="btnSmall btn "
              onClick={[Function]}
            >
              + det_reader
            </button>
            <button
              className="btnSmall btn "
              onClick={[Function]}
            >
              + fv_classify
            </button>
            <button
              className="btnSmall btn "
              onClick={[Function]}
            >
              + fv_cluster
            </button>
            <button
              className="btnSmall btn "
              onClick={[Function]}
            >
              + fv_sim
            </button>
            <button
              className="btnSmall btn "
              onClick={[Function]}
            >
              + join
            </button>
            <button
              className="btnSmall btn "
              onClick={[Function]}
            >
              + line_cross
            </button>
            <button
              className="btnSmall btn "
              onClick={[Function]}
            >
              + print
            </button>
            <button
              className="btnSmall btn "
              onClick={[Function]}
            >
              + SQL
            </button>
            <button
              className="btnSmall btn "
              onClick={[Function]}
            >
              + window
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="taskbarElement"
      style={
        Object {
          "right": "0px",
          "transform": "translate3d(0px, 0px, 0)",
          "width": "450px",
        }
      }
    >
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled={false}
        aria-roledescription="draggable"
        className="taskbarHeader"
        onKeyDown={[Function]}
        onPointerDown={[Function]}
        role="button"
        tabIndex={0}
        title="Drag Element"
      />
      <button
        className="minimiseButton"
        onClick={[Function]}
      >
        _
      </button>
      <div>
        <h3>
          Select a node to view it's properties
        </h3>
        <div>
          <h3>
            Current Nodes
          </h3>
          <br />
        </div>
      </div>
    </div>
    <div
      className="taskbarElement"
      style={
        Object {
          "top": "420px",
          "transform": "translate3d(0px, 0px, 0)",
          "width": "300px",
        }
      }
    >
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled={false}
        aria-roledescription="draggable"
        className="taskbarHeader"
        onKeyDown={[Function]}
        onPointerDown={[Function]}
        role="button"
        tabIndex={0}
        title="Drag Element"
      />
      <button
        className="minimiseButton"
        onClick={[Function]}
      >
        _
      </button>
      <div>
        <div
          className=""
        >
          <h3>
            Import & Export Graph
          </h3>
          <button
            className=" btn "
            onClick={[Function]}
          >
            Paste JSON / SVG code
          </button>
          <hr />
          <div
            className="btnGroup undefined"
          >
            <button
              className=" btn "
              onClick={[Function]}
            >
              Save JSON
            </button>
            <button
              className=" btn "
              onClick={[Function]}
            >
              Export SVG
            </button>
          </div>
          <br />
          <br />
          <label>
            <input
              checked={true}
              onChange={[Function]}
              type="checkbox"
            />
             Open in new window instead of downloading
          </label>
          <br />
          <br />
        </div>
      </div>
    </div>
    <div
      id="ddctx"
    />
    <div
      id="DndDescribedBy-0"
      style={
        Object {
          "display": "none",
        }
      }
    >
      
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
    </div>
    <div
      aria-atomic={true}
      aria-live="assertive"
      id="DndLiveRegion-0"
      role="status"
      style={
        Object {
          "border": 0,
          "clip": "rect(0 0 0 0)",
          "clipPath": "inset(100%)",
          "height": 1,
          "margin": -1,
          "overflow": "hidden",
          "padding": 0,
          "position": "fixed",
          "whiteSpace": "nowrap",
          "width": 1,
        }
      }
    />
  </div>
</div>
`;
